src/rviz/default_plugin/view_controllers/tf_aligned_view_controller.cpp

```diff
@@ -0,0 +1,128 @@
+ *       contributors may be used to endorse or promote products derived from
+ *       this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <OgreQuaternion.h>
+#include <OgreVector3.h>
+#include <OgreCamera.h>
+
+#include "rviz/viewport_mouse_event.h"
+#include "rviz/properties/bool_property.h"
+#include "rviz/properties/enum_property.h"
+#include "rviz/properties/float_property.h"
+
+#include "tf_aligned_view_controller.h"
+
+namespace rviz
+{
+
+TfAlignedViewController::TfAlignedViewController()
+{
+  axis_property_ = new EnumProperty("Align to", "z-axis", "Align the camera to this axis.", this);
+  axis_property_->addOptionStd("x-axis", 1);
+  axis_property_->addOptionStd("y-axis", 2);
+  axis_property_->addOptionStd("z-axis", 3);
+
+  invert_axis_ = new BoolProperty("Invert Axis", true, "Align camera with the negative axis.", this);
+  camera_roll_property_ = new FloatProperty("Camera Roll", 0, "Roll about the camera's view axis.", this);
+}
+
+TfAlignedViewController::~TfAlignedViewController()
+{
+}
+
+void TfAlignedViewController::onInitialize()
+{
+  FramePositionTrackingViewController::onInitialize();
+  camera_->setProjectionType( Ogre::PT_PERSPECTIVE );
+  invert_z_->hide();
+}
+
+void TfAlignedViewController::reset()
+{
+  camera_->setOrientation(reference_orientation_);
+  camera_->setPosition(reference_position_);
+  updateCamera();
+}
+
+void TfAlignedViewController::update(float dt, float ros_dt)
+{
+  FramePositionTrackingViewController::update( dt, ros_dt );
+  updateCamera();
+}
+
+void TfAlignedViewController::onTargetFrameChanged(const Ogre::Vector3& /* old_reference_position */,
+                                                   const Ogre::Quaternion& /* old_reference_orientation */)
+{
+  camera_->setOrientation(reference_orientation_);
+  camera_->setPosition(reference_position_);
+}
+
+void TfAlignedViewController::updateCamera()
+{
+  Ogre::Quaternion q_offset;
+  int view_direction = invert_axis_->getBool() ? -1 : 1;
+  int view_axis = axis_property_->getOptionInt() * view_direction;
+
+  // reference_orientation_ will be aligned to -z-axis by default.
+  switch (view_axis) {
```

Comment:
Using hand-crafted equations for these computations is bad programming practice: The source code doesn't reveal the underlying concepts, rendering this code nearly un-maintainable.